File: app/src/main/java/com/cookplan/models/RatioMeasure.java
package com.cookplan.models;

/**
 * Created by DariaEfimova on 06.04.17.
 */

import android.os.Parcel;
import android.os.Parcelable;

/**
 * Class for saving the ratio between 2 measure units
 * <p>
 * rightUnit = ratio * leftUnit
 */

public class RatioMeasure implements Parcelable {

    public Double ratio;
    public MeasureUnit rightUnit;
    public MeasureUnit leftUnit;

    public RatioMeasure() {
    }

    public RatioMeasure(MeasureUnit rightUnit, Double ratio, MeasureUnit leftUnit) {
        this();
        this.ratio = ratio;
        this.rightUnit = rightUnit;
        this.leftUnit = leftUnit;
    }

    public Double getRatio() {
        return ratio;
    }

    public void setRatio(Double ratio) {
        this.ratio = ratio;
    }

    public MeasureUnit getRightUnit() {
        return rightUnit;
    }

    public void setRightUnit(MeasureUnit rightUnit) {
        this.rightUnit = rightUnit;
    }

    public MeasureUnit getLeftUnit() {
        return leftUnit;
    }

    public void setLeftUnit(MeasureUnit leftUnit) {
        this.leftUnit = leftUnit;
    }

    public Double getMultiplier(MeasureUnit fromUnit, MeasureUnit toUnit) {
        if (fromUnit == rightUnit && toUnit == leftUnit) {
            return ratio;
        }
        if (fromUnit == leftUnit && toUnit == rightUnit) {
            return 1 / ratio;
        }
        return -1.;
    }

    // Parcelling part
    public RatioMeasure(Parcel in) {
        ratio = in.readDouble();
        rightUnit = (MeasureUnit) in.readSerializable();
        leftUnit = (MeasureUnit) in.readSerializable();
    }

    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeDouble(ratio);
        dest.writeSerializable(rightUnit);
        dest.writeSerializable(leftUnit);
    }

    public static final Parcelable.Creator CREATOR = new Parcelable.Creator() {
        public RatioMeasure createFromParcel(Parcel in) {
            return new RatioMeasure(in);
        }

        public RatioMeasure[] newArray(int size) {
            return new RatioMeasure[size];
        }
    };
}

File: app/src/main/java/com/cookplan/models/RatioMeasure.kt
package com.cookplan.models

/**
 * Created by DariaEfimova on 06.04.17.
 */

import android.os.Parcel
import android.os.Parcelable

/**
 * Class for saving the ratio between 2 measure units
 *
 *
 * rightUnit = ratio * leftUnit
 */

data class RatioMeasure(var ratio: Double = 0.toDouble(),
                        var rightUnit: MeasureUnit,
                        var leftUnit: MeasureUnit) : Parcelable {
    constructor() : this(1.toDouble(), MeasureUnit.UNITS, MeasureUnit.UNITS)

    fun getMultiplier(fromUnit: MeasureUnit, toUnit: MeasureUnit): Double {
        if (fromUnit === rightUnit && toUnit === leftUnit) {
            return ratio
        }
        return if (fromUnit === leftUnit && toUnit === rightUnit) {
            1 / ratio
        } else -1.0
    }

    constructor(source: Parcel) : this(
            source.readValue(Double::class.java.classLoader) as Double,
            source.readSerializable() as MeasureUnit,
            source.readSerializable() as MeasureUnit)

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeValue(ratio)
        writeSerializable(rightUnit?.ordinal)
        writeSerializable(leftUnit?.ordinal)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<RatioMeasure> = object : Parcelable.Creator<RatioMeasure> {
            override fun createFromParcel(source: Parcel): RatioMeasure = RatioMeasure(source)
            override fun newArray(size: Int): Array<RatioMeasure?> = arrayOfNulls(size)
        }
    }
}
